### ISSUE TITLE
The role of the hyper param Beta

### ISSUE THREAD

**Issue by Saltychtao** [eric-mitchell/direct-preference-optimization#4]
Hello,
I notice that both in your paper and the implementation, the hyper param beta is just a scale of the loss function, rather than a balancing factor between two variables. Can we just ignore the beta hyper param?

Thanks

**Comment by eric-mitchell** [comment]
It's not accurate to say that beta just scales the DPO loss. It scales the _pre-sigmoid_ reward difference (see [here](https://github.com/eric-mitchell/direct-preference-optimization/blob/main/trainers.py#L73)).

<img width="943" alt="Screenshot 2023-06-27 at 12 55 46 AM" src="https://github.com/eric-mitchell/direct-preference-optimization/assets/56408839/5f40d2e4-cf83-4b7f-93ca-2d5d45338f17">

If we use very small beta, we basically only care about assigning much higher probability to the chosen response than the rejected response, and if beta is too small, the model can eventually over-optimize reward and degenerate (see [here](https://arxiv.org/abs/2210.10760)). With very high beta, we'll have a stronger constraint to keep the policy we're learning close to the reference model.


TL;DR: beta controls how strong our KL regularization toward the reference model is. You definitely can't ignore it!

Closing, but feel free to re-open if this answer doesn't resolve your question.

**Comment by Saltychtao** [comment]
Thanks for your quick response, however, I still don't get it. The beta is multiplied to both the reward of the chosen response and the rejected response. Therefore, should not the effects of a small beta on the probability of chosen and rejected responses be the same?

I plot the function `-log sigmoid (\beta x)`, and vary the value of beta.
![image](https://github.com/eric-mitchell/direct-preference-optimization/assets/9932507/77a94afe-43ab-4ccd-bce3-0ee91e467dea)

It can be observed that \beta does have an effect on the loss function instead of a scaling factor. When \beta is small, the function curve is flatter, suggesting we are treating small and large reward estimation errors are optimized more equally. However, I cannot establish a relationship between this behavior and the role of \beta in the original KL divergency in PPO loss.

Can you please explain more about this? Thanks.

**Comment by eric-mitchell** [comment]
To see how your observation about beta flattening the loss curve $\ell = -\log \frac{1}{1 + \exp (-\beta x)}$, it's helpful to think about what $x$ is in the case of the DPO loss. For the DPO loss, we have $x = \log \frac{p_\theta(y_w \mid x)}{p_{ref}(y_w \mid x)} - \log \frac{p_\theta(y_l \mid x)}{p_{ref}(y_l \mid x)}$, where the first term is measures "does our policy assign higher likelihood to the chosen/preferred response than the reference model assigns to the chosen/preferred response, given the prompt?" Call this the "chosen margin" $c = \log \frac{p_\theta(y_w \mid x)}{p_{ref}(y_w \mid x)}$. The second term measures "does our policy assign higher likelihood to the rejected/dispreferred response than the reference model assigns to the rejected/dispreferred response, given the prompt?" Call this the "rejected margin" $r = \log \frac{p_\theta(y_l \mid x)}{p_{ref}(y_l \mid x)}$.

The goal of DPO, roughly speaking, is to make the chosen margin larger than the rejected margin; hopefully this is intuitive, as we want our policy's relative preference (compared to the reference model) for the chosen completion to be larger than its relative preference for the rejected response (again, compared to the reference model).

Now let's think about what it means to say $x = c - r$, with respect to your plots. As you point out, making beta smaller makes the curve flatter. In other words, in order to achieve some low loss $\epsilon$, $x$ must be relatively large. Since $x = c - r$, this means that our policy will have to make $c$ very large relative to $r$, i.e., the policy will have to assign _much_ higher probability to the chosen completion than the reference model, so the policy will diverge from the reference model significantly (it is also possible to just assign much lower probability to the rejected completion than the reference, but the argument is the same). If beta is large, then the loss gets very small even when $x$ is only slightly larger than zero, that is, our policy only needs to prefer the chosen completion a _little bit_ more than the reference model in order for the loss to nearly converge. Thus, when beta is larger, we essentially stop optimizing the policy once it's drifted just a little bit from the reference model.

**Comment by Saltychtao** [comment]
I understand now! Thank you very much for your detailed explanation!

**Comment by Saltychtao** [comment]
Sorry to bother you, but I have another quesstion.

In my use case, I am trying to use DPO to optimize a Chinese-English translation model to fix the off-target problem, i.e. English translations contain some Chinese characters. After setting beta to a relatively large value(0.8 - 1), the model's loss quickly coverages to near 0, which makes sense according to our earlier discussion. However, after inspecting the generation results, I still find them degenerate, containing non-English-non-Chinese words and repeated words. 

Can you please provide some suggestions on this? Do you think adding an additional LM loss would help?

Thanks in advance.

**Comment by eric-mitchell** [comment]
Happy to debug this- can you open a new issue to discuss in? In the new issue, can you share the wandb logs for your runs? That will help me debug. Also, did you run SFT first? And are you using a public preference/SFT dataset?

With beta 1, my very rough intuition is that if anything, the policy would be too strongly constrained to stay close to the reference model.
